### PR TITLE
Replaces KVHardcodedDictionary with dynamic version

### PR DESCRIPTION
**Pull request by slovit**

### CODE REVIEW

**Comment by slovit** [comment]
test this please

File: dcc-submission-server/src/main/java/org/icgc/dcc/submission/service/MailService.java [hunk 1]
```diff
@@ -92,7 +92,7 @@ public void sendSupportFeedback(@NonNull Feedback feedback) {
 
     send(
         feedback.getEmail(),
-        properties.getMail().getSupportEmail(),
+        properties.getEmail().getSupportEmail(),
```
Comment (btiernay) [inline dcc-submission-server/src/main/java/org/icgc/dcc/submission/service/MailService.java]: We should probably undo this.

Comment (slovit) [reply dcc-submission-server/src/main/java/org/icgc/dcc/submission/service/MailService.java]: Okay, but the build will fail. `MAIL` is unset on all the slaves.

Comment (btiernay) [reply dcc-submission-server/src/main/java/org/icgc/dcc/submission/service/MailService.java]: I know, that's why I sent you a slack request a few days ago re: this.

Comment (slovit) [reply dcc-submission-server/src/main/java/org/icgc/dcc/submission/service/MailService.java]: I know you sent a request and I unset `MAIL` on all the slaves as you asked, restarted all the slaves and it still fails.

File: dcc-submission-core/src/main/java/org/icgc/dcc/submission/dictionary/model/FileSchema.java [hunk 2]
```diff
@@ -336,4 +333,40 @@ public boolean apply(Field field) {
...
   }
 
+  /**
+   * Compares according to {@link Relation}. If {@code this} is a parent of the {@code ot}
+   * @return -1 if {@code this} is a parent of the {@code other}<br>
+   * 0 if they are equal<br>
+   * 1 if {@code this} is a child of the {@code other}
+   */
+  @Override
+  public int compareTo(FileSchema other) {
```
Comment (btiernay) [inline dcc-submission-core/src/main/java/org/icgc/dcc/submission/dictionary/model/FileSchema.java]: I'm not sure `Comparable` is the right mechanism here. This is a very context specific ordering that probably would be better served in a `DictionaryTopologicalComparator` implementation.

http://stackoverflow.com/questions/2266827/when-to-use-comparable-and-comparator

File: dcc-submission-core/src/test/resources/logback-test.xml [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+-->
+<configuration debug="false">
+  <property name="log.dir" value="target/logs" />
+  <property name="log.name" value="dcc-server" />
+  <property name="log.debug.name" value="dcc-server.debug" />
```
Comment (btiernay) [inline dcc-submission-core/src/test/resources/logback-test.xml]: Please update

File: dcc-submission-core/src/test/resources/logback-test.xml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    <File>${log.dir}/${log.debug.name}.log</File>
+    <rollingPolicy class="ch.qos.logback.core.rolling.TimeBasedRollingPolicy">
+      <FileNamePattern>${log.dir}/${log.debug.name}.%d{yyyy-MM-dd}.log</FileNamePattern>
+    </rollingPolicy>
+    <encoder>
+      <pattern>${pattern}</pattern>
+    </encoder>
+  </appender>
+
+  <logger name="org.icgc.dcc.submission.core.UserService" level="info" additivity="false">
```
Comment (btiernay) [inline dcc-submission-core/src/test/resources/logback-test.xml]: None of these exist

File: dcc-submission-core/src/test/resources/logback-test.xml [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  </logger>
+  <logger name="org.icgc.dcc.submission.http.jersey.BasicHttpAuthenticationRequestFilter" level="info" additivity="false">
+    <appender-ref ref="DEBUG" />
+  </logger>
+  <logger name="org.apache.sshd.server.session.ServerSession" level="info" additivity="false">
+    <appender-ref ref="DEBUG" />
+  </logger>
+
+  <logger name="de.flapdoodle.embed" level="WARN" />
+  <logger name="org.glassfish.jersey" level="WARN" />
```
Comment (btiernay) [inline dcc-submission-core/src/test/resources/logback-test.xml]: Doesn't exist

File: dcc-submission-test/pom.xml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    <!-- Utilities -->
+    <dependency>
+      <groupId>org.projectlombok</groupId>
+      <artifactId>lombok</artifactId>
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>${junit.version}</version>
```
Comment (btiernay) [inline dcc-submission-test/pom.xml:52]: This version tag needed?

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java [hunk 7]
```diff
@@ -17,156 +17,342 @@
...
+    // Primary keys
+    val builder = ImmutableMap.<KVKeyType, Multimap<KVFileType, Integer>> builder();
+    val pks = getKeysMultimap();
+    val primaryKeys = getPrimaryKeyNames(fileType);
+    pks.putAll(fileType, resolveKeyIndices(schemaFieldNames, primaryKeys));
+    builder.put(PK, pks);
 
-        }));
+    val relations = fileSchema.getRelations();
+    // We assumed that optionals mean which fields should be used to relate to the other file type.
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java:73]: 👍

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java [hunk 8]
```diff
@@ -17,156 +17,342 @@
...
-
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getOptionalReferencedFileType2(org.icgc.dcc.submission
-   * .validation.key.core.KVFileType)
-   */
+  public List<String> getErrorFieldNames(@NonNull KVFileType fileType, @NonNull KVErrorType errorType,
+      @NonNull Optional<KVFileType> optionalReferencedFileType) {
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java]: With respect to `Optional` parameters, something to consider:

http://blog.jhades.org/java-8-how-to-use-optional/
http://stackoverflow.com/questions/31922866/why-should-java-8s-optional-not-be-used-in-arguments

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java [hunk 9]
```diff
@@ -17,156 +17,342 @@
...
-
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getPrimaryKeyNames(org.icgc.dcc.submission.validation.
-   * key.core.KVFileType)
-   */
+  public Iterable<KVFileType> getTopologicallyOrderedFileTypes() {
+    return KVDynamicDictionaryHelper.getTopologicallyOrderedFileTypes(dictionary);
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java]: Not a fan of `Helper` classes as it is a `Utils` in general and thus typically has poor cohesion.

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java [hunk 10]
```diff
@@ -17,156 +17,342 @@
...
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getReferencingFileType(org.icgc.dcc.submission.validation
-   * .key.core.KVFileType)
-   */
+  public boolean hasChildren(@NonNull KVFileType fileType) {
+    return dictionary.getFiles().stream()
+        .flatMap(file -> file.getRelations().stream())
+        .map(relation -> relation.getOtherFileType())
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java]: `Foo::getOtherFileType`?

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java [hunk 11]
```diff
@@ -17,156 +17,342 @@
...
-  /*
-   * (non-Javadoc)
-   * 
-   * @see
-   * org.icgc.dcc.submission.validation.key.core.KVDictionary#getReferencingFileType(org.icgc.dcc.submission.validation
-   * .key.core.KVFileType)
-   */
+  public boolean hasChildren(@NonNull KVFileType fileType) {
+    return dictionary.getFiles().stream()
+        .flatMap(file -> file.getRelations().stream())
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java:183]: FYI, could also do:

``` java
.map(FileSchema::getRelations())
.flatMap(Collection::stream)
```

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java [hunk 12]
```diff
@@ -17,156 +17,342 @@
...
+        .map(relation -> relation.getOtherFileType())
+        .map(KVFileType::from)
+        .anyMatch(ft -> ft == fileType);
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java]: Could use:
https://docs.oracle.com/javase/8/docs/api/java/util/function/Predicate.html#isEqual-java.lang.Object-

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java [hunk 13]
```diff
@@ -17,156 +17,342 @@
...
+        .map(KVFileType::from)
+        .collect(toImmutableList());
+  }
+
+  private Map<KVFileType, RowConditionEvaluator> createConditionEvaluators(@NonNull KVFileType fileType) {
+    val conditions = resolveConditions(fileType);
+    val conditionFields = getTypeFields(fileType, dictionary);
+
+    return conditions.entrySet().stream()
+        .collect(toImmutableMap(entry -> entry.getKey(),
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java:205]: There is another version of `toImmutableMap` that doesn't require this first argument as it is almost always this pattern.

File: dcc-submission-validator/src/test/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionaryTest.java [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+    assertKey(keys, PK, ImmutableMap.of(BIOMARKER, ImmutableList.of(2, 0, 1)));
+    assertKey(keys, FK, ImmutableMap.of(DONOR, ImmutableList.of(0), SPECIMEN, ImmutableList.of(1)));
+    assertKey(keys, OPTIONAL_FK, emptyMap());
+    assertKey(keys, CONDITIONAL_FK, emptyMap());
+  }
+
+  @Test
+  public void testGetTopologicallyOrderedFileTypes() throws Exception {
+    ImmutableList.copyOf(kvDictionary.getTopologicallyOrderedFileTypes());
+    // TODO: Test for order
```
Comment (btiernay) [inline dcc-submission-validator/src/test/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionaryTest.java]: Probably a good idea no?

**Comment by btiernay** [comment]
A small feature we probably want to add is the ability to have access to the project code. This is currently available in the ScriptRestriction:

https://github.com/icgc-dcc/dcc-submission/blob/develop/dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/primary/restriction/ScriptRestriction.java#L361

File: dcc-submission-core/src/test/java/org/icgc/dcc/submission/dictionary/model/DictionaryTest.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+import com.google.common.collect.Lists;
+
+public class DictionaryTest extends AbstractDictionaryTest {
+
+  Dictionary dictionary;
+
+  @Override
+  @SneakyThrows
+  public void setUp() {
+    super.setUp();
```
Comment (btiernay) [inline dcc-submission-core/src/test/java/org/icgc/dcc/submission/dictionary/model/DictionaryTest.java]: http://beust.com/weblog2/archives/000252.html

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/RowConditionEvaluator.java [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+public class RowConditionEvaluator {
+
+  private final ExecutableStatement statement;
+  private final List<String> fieldNames;
+
+  public RowConditionEvaluator(@NonNull String script, @NonNull List<String> fieldNames) {
+    this.statement = (ExecutableStatement) MVEL.compileExpression(script);
+    this.fieldNames = fieldNames;
+
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/RowConditionEvaluator.java]: Remove new line.

File: dcc-submission-test/pom.xml (hunk 6 above)
Comment (slovit) [reply dcc-submission-test/pom.xml:52]: Yes, otherwise the dependency is not resolvable

Comment (btiernay) [reply dcc-submission-test/pom.xml:52]: https://github.com/icgc-dcc/dcc-parent/blob/develop/pom.xml#L703

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java (hunk 9 above)
Comment (slovit) [reply dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java]: How you want it name it?

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java (hunk 11 above)
Comment (slovit) [reply dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java:183]: 👍 
But I like my version more as it encodes the whole intent in one line

```
.flatMap(Collection::stream)
```

seems a supportive operation

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionaryHelper.java [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+
+import org.icgc.dcc.submission.dictionary.model.Dictionary;
+import org.icgc.dcc.submission.dictionary.model.FileSchema;
+import org.jgrapht.experimental.dag.DirectedAcyclicGraph;
+import org.jgrapht.graph.DefaultEdge;
+
+import com.google.common.collect.ImmutableList;
+
+@NoArgsConstructor(access = PRIVATE)
+public final class KVDynamicDictionaryHelper {
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionaryHelper.java]: `KVFileTypeRelations`

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionaryHelper.java [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+    val fileType = KVFileType.from(schema.getFileType());
+    val parents = schema.getRelations().stream()
+        .map(relation -> relation.getOtherFileType())
+        .map(KVFileType::from)
+        .collect(toImmutableList());
+
+    return new KVFileTypeRelation(fileType, parents);
+  }
+
+  public static class Edge extends DefaultEdge {
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionaryHelper.java]: `private`?

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java (hunk 13 above)
Comment (slovit) [reply dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java:205]: It seems the other function consumes a `keyMapper`, but I need a `valueMapper` here

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionaryHelper.java [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+    val graph = new DirectedAcyclicGraph<KVFileType, Edge>(Edge.class);
+
+    fileTypeRelations.forEach(fileTypeRelation -> graph.addVertex(fileTypeRelation.getFileType()));
+
+    fileTypeRelations.forEach(fileTypeRelation -> {
+      KVFileType child = fileTypeRelation.fileType;
+      fileTypeRelation.parents.stream()
+          .forEach(parent -> {
+            try {
+              graph.addDagEdge(parent, child);
```
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionaryHelper.java]: https://github.com/jOOQ/jOOL#orgjooqlambdaunchecked . Might be nice to add to our common `Consumers` class.

File: dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java (hunk 13 above)
Comment (btiernay) [inline dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java]:
```java
  public static <T, K, V> Collector<T, ImmutableMap.Builder<K, V>, ImmutableMap<K, V>> toImmutableMap(
      @NonNull Function<? super T, ? extends K> keyMapper, @NonNull Function<? super T, ? extends V> valueMapper) {

    final BiConsumer<ImmutableMap.Builder<K, V>, T> accumulator =
        (builder, entry) -> builder.put(keyMapper.apply(entry), valueMapper.apply(entry));

    return Collector.of(
        ImmutableMap.Builder::new,
        accumulator,
        (b1, b2) -> b1.putAll(b2.build()),
        ImmutableMap.Builder::build);
  }
```

Comment (btiernay) [reply dcc-submission-validator/src/main/java/org/icgc/dcc/submission/validation/key/core/KVDynamicDictionary.java:205]: You are right. My mistake.

**Pull request merged**